### PR TITLE
[JENKINS-70044] Add telemetry for activation of optional permissions

### PR DESCRIPTION
**Pull request by dwnusbaum**
See [JENKINS-70044](https://issues.jenkins.io/browse/JENKINS-70044).

Note that we would like to backport this change to the LTS. https://www.jenkins.io/download/lts/#backporting-process does not list telemetry explicitly, but otherwise we will have to set activation dates far into the future and then wait for this to be picked up by the next LTS baseline. I recall @daniel-beck mentioning that new telemetry collectors have been backported in the past, but I could not find any discussion of this from a very quick search on the developer mailing list.

### Testing done

I tested this PR by replacing [`Telemetry.ENDPOINT`](https://github.com/jenkinsci/jenkins/blob/5bf632f261bacb214b2ba499890c5ba3ffea2189/core/src/main/java/jenkins/telemetry/Telemetry.java#L75) with a link to https://webhook.site, then I called `ExtensionList.lookupSingleton(jenkins.telemetry.impl.OptionalPermissions.class).run()` from the Jenkins script console and verified that the request showed up with the expected content. I then installed `credentials` and `manage-permission` from the Plugin Manager, reran the script, and verified that the credentials-related permissions were included in the payload and that `hudson.model.Hudson.Manage` was now marked as being enabled.

<details>
<summary>Sample payload</summary>

```
{
    "components": {
        "bouncycastle-api": "2.26",
        "credentials": "1189.vf61b_a_5e2f62e",
        "instance-identity": "116.vf8f487400980",
        "javax-activation-api": "1.2.0-5",
        "javax-mail-api": "1.6.2-8",
        "jenkins-core": "2.378-SNAPSHOT",
        "manage-permission": "1.0.1",
        "mina-sshd-api-common": "2.9.1-44.v476733c11f82",
        "mina-sshd-api-core": "2.9.1-44.v476733c11f82",
        "sshd": "3.249.v2dc2ea_416e33",
        "structs": "324.va_f5d6774f3a_d"
    },
    "permissions": {
        "com.cloudbees.plugins.credentials.CredentialsProvider.UseItem": false,
        "com.cloudbees.plugins.credentials.CredentialsProvider.UseOwn": false,
        "hudson.model.Computer.ExtendedRead": false,
        "hudson.model.Hudson.Manage": true,
        "hudson.model.Hudson.SystemRead": false,
        "hudson.model.Item.ExtendedRead": false,
        "hudson.model.Item.WipeOut": false,
        "hudson.model.Run.Artifacts": false
    }
}
```
</details>

I did not add any automated tests. [`TelemetryTest`](https://github.com/jenkinsci/jenkins/blob/5bf632f261bacb214b2ba499890c5ba3ffea2189/test/src/test/java/jenkins/telemetry/TelemetryTest.java) already covers the overall behavior of the `Telemetry` extension point. I can add a test that checks the output of `OptionalPermissions.createContent` if reviewers think that would be useful.

### Proposed changelog entries

- Add telemetry for activation of permissions that are not enabled by default.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7342"></a>

### CODE REVIEW

File: core/src/main/java/jenkins/telemetry/impl/OptionalPermissions.java
Comment (dwnusbaum) [inline core/src/main/java/jenkins/telemetry/impl/OptionalPermissions.java:73]: I am not sure what dates make the most sense. I set the start to 2022-11-01 to facilitate testing. Depending on whether we can backport this and what the LTS schedule is like, maybe `2023-02-01` could be a good end date, but my concern is that many users will refrain from updating during the holiday season and so December and January data would be sparse and perhaps non-representative of the overall userbase.

File: core/src/main/resources/jenkins/telemetry/impl/OptionalPermissions/description.jelly
Comment (dwnusbaum) [inline core/src/main/resources/jenkins/telemetry/impl/OptionalPermissions/description.jelly:5]: The phrasing here seems kind of awkward; suggestions welcome.

**Comment by daniel-beck** [comment]
> I recall @daniel-beck mentioning that new telemetry collectors have been backported in the past, but I could not find any discussion of this from a very quick search on the developer mailing list.

Some past examples I was involved with, basically everything has been backported into LTS:

[JENKINS-53768](https://issues.jenkins.io/browse/JENKINS-53768) backported initial telemetry infrastructure into 2.138.2 with `@Restricted` (discussed and approved in the 2018-09-26 project meeting).
[JENKINS-54029](https://issues.jenkins.io/browse/JENKINS-54029) backported Stapler dispatch telemetry into 2.138.3.
[JENKINS-57167](https://issues.jenkins.io/browse/JENKINS-57167) backported re-activation of Stapler telemetry into 2.164.3.
[JENKINS-67229](https://issues.jenkins.io/browse/JENKINS-67229) backported agent-to-controller file path access logging telemetry into 2.319.1.

I don't know whether there's been an "official" discussion on the topic of backporting eligibility of telemetry. In the initial JEP request for comments thread I wrote,

> It's also difficult to say right now how we'd handle LTS releases -- would collectors be eligible for backports? (I say yes, as optional extension points written such that they mostly silently fail.) 

but there were no further responses on the topic (at least no active rejections of this).

To me, the backporting requirement basically follows logically from the idea of keeping the collection period short, ideally just a few months to collect less data. Anything less than around 4-6 months or so given the would not result in a broad user base to collect from, making data less meaningful, and specifying very long durations to cover both LTS and weekly will result in the unnecessary collection of data from weekly users. We could specify dates several months into the future, but, aside from causing delays in the projects these collectors would support, that may also cause bugs in the implementation (and we've had those) to only be discovered months after the code has been added to all recent releases, both weekly and LTS, removing our ability to quickly react to regressions reported in weekly releases.

And many past implementations collected data for just this short of a period, producing useful data during that time:
* `security-system-properties` (the JEP-214 reference impl) collected for 3 months,
* AcceptLanguages for 3 months,
* Stapler (1): ~3.7 months (the re-activation added ~3 months),
* agent-to-controller file path for just over 3 months once it was merged.

None of these would have been able to cover data from across the user base (i.e., including LTS) without a backport.

File: core/src/main/resources/jenkins/telemetry/impl/OptionalPermissions/description.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/telemetry/impl/OptionalPermissions/description.jelly:4]:
```suggestion
    This trial collects which of these permissions are enabled and disabled, as well as the list of installed plugins.
```

**Review by daniel-beck — APPROVED** [review]
Looks good and seems to contain all of the disabled-by-default permissions.

Suggested docs change to make this accurate (we don't actually collect system properties).

**Review by Wadeck — APPROVED** [review]
👍

**Review by NotMyFault — APPROVED** [review]
Fine with me to backport this into 2.375.1 👍🏻

**Review by yaroslavafenkin — APPROVED** [review]

**Comment by NotMyFault** [comment]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Pull request merged**
